optaplanner-examples/src/main/java/org/optaplanner/examples/pas/solver/score/PatientAdmissionMoveConstraintProvider.java

```diff
@@ -0,0 +1,180 @@
+package org.optaplanner.examples.pas.solver.score;
+
+import org.optaplanner.core.api.score.buildin.hardmediumsoft.HardMediumSoftScore;
+import org.optaplanner.core.api.score.stream.Constraint;
+import org.optaplanner.core.api.score.stream.ConstraintFactory;
+import org.optaplanner.core.api.score.stream.ConstraintProvider;
+import org.optaplanner.examples.pas.domain.BedDesignation;
+import org.optaplanner.examples.pas.domain.Department;
+import org.optaplanner.examples.pas.domain.DepartmentSpecialism;
+import org.optaplanner.examples.pas.domain.Gender;
+import org.optaplanner.examples.pas.domain.GenderLimitation;
+import org.optaplanner.examples.pas.domain.PreferredPatientEquipment;
+import org.optaplanner.examples.pas.domain.RequiredPatientEquipment;
+import org.optaplanner.examples.pas.domain.RoomEquipment;
+import org.optaplanner.examples.pas.domain.RoomSpecialism;
+
+import static org.optaplanner.core.api.score.stream.Joiners.equal;
+import static org.optaplanner.core.api.score.stream.Joiners.filtering;
+import static org.optaplanner.core.api.score.stream.Joiners.lessThan;
+
+/*
+ * This is constraints for Hospital Bed Planning
+ * They are based on patientAdmissionScheduleConstraints.drl
+ * Planning Entity: BedDesignation
+ * Planning Variable: Bed(nullable) - would not be prefiltered on uninitialized solutions
+ * Bed is nullable so in case you need to access it members check that planning value bed is not null
+ */
+
+public class PatientAdmissionMoveConstraintProvider implements ConstraintProvider {
+
+    @Override
+    public Constraint[] defineConstraints(ConstraintFactory constraintFactory) {
+        return new Constraint[]{
+                sameBedInSameNightConstraint(constraintFactory),
+                femaleInMaleRoomConstraint(constraintFactory),
+                maleInFemaleRoomConstraint(constraintFactory),
+                differentGenderInSameGenderRoomInSameNightConstraint(constraintFactory),
+                departmentMinimumAgeConstraint(constraintFactory),
+                departmentMaximumAgeConstraint(constraintFactory),
+                requiredPatientEquipmentConstraint(constraintFactory),
+                assignEveryPatientToABedConstraint(constraintFactory),
+                preferredMaximumRoomCapacityConstraint(constraintFactory),
+                departmentSpecialismConstraint(constraintFactory),
+                roomSpecialismNotExistsConstraint(constraintFactory),
+                roomSpecialismNotFirstPriorityConstraint(constraintFactory),
+                preferredPatientEquipmentConstraint(constraintFactory)
+        };
+    }
+
+    public Constraint sameBedInSameNightConstraint(ConstraintFactory constraintFactory) {
+        return constraintFactory.fromUniquePair(BedDesignation.class)
+                .filter((left, right) -> left.getBed() != null
+                        && right.getBed() != null
+                        && left.getBed() == right.getBed()
```

Comment:
Please use `Joiners.equal(BedDesignation::getBed)` for this.

Reply:
BTW is that worth to mention that `fromUniquePair` is a syntax sugar as here: http://github.com/dupliaka/optaplanner/blob/PLANNER-1823/optaplanner-core/src/main/java/org/optaplanner/core/api/score/stream/ConstraintFactory.java#L95-L95
Since I was beliving that sugared syntax should behave absolutely the same under the hood.

Reply:
I mean since fromUniquePair implemented in the way that it is much slower

Reply:
I believe that the Javadoc is correct - since `join`, indeed, also filters out uninitialized entities.

Reply:
+I have a concern about that: if I will use regular join, then since bed is nullable I need to check both (left and right) streams for bed!=null, but I will have a problem there bcs I need to do that before joining and I can not do that

Reply:
```    public Constraint sameBedInSameNightConstraint(ConstraintFactory constraintFactory) {
        return constraintFactory.from(BedDesignation.class)
                .filter((left)->left.getBed()!=null) // cannot filter right
                .join(BedDesignation.class,
                        equal(BedDesignation::getBed),
                        lessThan(BedDesignation::getId))
                .penalize("sameBedInSameNight", HardMediumSoftScore.ofHard(1000)));
    }```

Reply:
That's the point though, you don't need to do that. The `bed != null` check is already done for you. This will be enough:

     return constraintFactory.from(BedDesignation.class)
                .join(BedDesignation.class,
                        equal(BedDesignation::getBed),
                        lessThan(BedDesignation::getId))
                .penalize("sameBedInSameNight", HardMediumSoftScore.ofHard(1000)));

In fact, now that this constraint is so simple, you probably just need:

     return constraintFactory.fromUniquePair(BedDesignation.class,
                equal(BedDesignation::getBed))
                .penalize("sameBedInSameNight", HardMediumSoftScore.ofHard(1000)));

Reply:
Brilliant idea

Reply:
https://github.com/kiegroup/optaplanner/pull/872/commits/7dd155181c6f1f7f867667d82eb53de3aa1d5a38